src/compiler/checker.ts

```diff
@@ -16991,31 +16986,43 @@ namespace ts {
             return false;
         }
 
-        // Return true if type is of the given kind. A union type is of a given kind if all constituent types
-        // are of the given kind. An intersection type is of a given kind if at least one constituent type is
-        // of the given kind.
-        function isTypeOfKind(type: Type, kind: TypeFlags): boolean {
-            if (type.flags & kind) {
+        function isTypeOfKind(source: Type, kind: TypeFlags, excludeAny?: boolean) {
+            if (source.flags & kind) {
                 return true;
             }
-            if (type.flags & TypeFlags.Union) {
-                const types = (<UnionOrIntersectionType>type).types;
-                for (const t of types) {
-                    if (!isTypeOfKind(t, kind)) {
-                        return false;
-                    }
-                }
-                return true;
+            if (excludeAny && source.flags & (TypeFlags.Any | TypeFlags.Void | TypeFlags.Undefined | TypeFlags.Null)) {
+                // TODO: The callers who want this should really handle these cases FIRST
+                return false;
             }
-            if (type.flags & TypeFlags.Intersection) {
-                const types = (<UnionOrIntersectionType>type).types;
-                for (const t of types) {
-                    if (isTypeOfKind(t, kind)) {
-                        return true;
-                    }
-                }
+            const targets = [];

```

Comment:
Why are you building up a union type here? It causes allocations and ends up being broken down to discrete checks for each contained type anyway. I think it would be better to just have a sequence of conditional `isTypeAssignableTo` calls here, e.g.

```ts
return kind & TypeFlags.NumberLike && isTypeAssignableTo(source, numberType) ||
    kind & TypeFlags.StringLike && isTypeAssignableTo(source, stringType) ||
    kind & TypeFlags.BooleanLike && isTypeAssignableTo(source, BooleanType) ||
    ...
```